File: libs/ontology-kernel/src/models/object_type_binding.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sqlx::FromRow;
use uuid::Uuid;

/// Sync mode for an `ObjectType`-to-dataset binding.
///
/// * `Snapshot` – every materialisation truncates the prior set of bound
///   instances and re-inserts the dataset rows.
/// * `Incremental` – upsert by primary key column (insert new, update existing).
/// * `View` – metadata-only binding; rows are not materialised, the read-side
///   is expected to serve them lazily from the dataset.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "text", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum ObjectTypeBindingSyncMode {
    Snapshot,
    Incremental,
    View,
}

impl ObjectTypeBindingSyncMode {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Snapshot => "snapshot",
            Self::Incremental => "incremental",
            Self::View => "view",
        }
    }
}

impl TryFrom<&str> for ObjectTypeBindingSyncMode {
    type Error = String;

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value.trim() {
            "snapshot" => Ok(Self::Snapshot),
            "incremental" => Ok(Self::Incremental),
            "view" => Ok(Self::View),
            other => Err(format!(
                "sync_mode '{other}' is not supported; expected one of: snapshot, incremental, view"
            )),
        }
    }
}

/// One source-to-property mapping inside a binding.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ObjectTypeBindingPropertyMapping {
    /// Column name in the source dataset row.
    pub source_field: String,
    /// Target property name on the ObjectType.
    pub target_property: String,
}

/// Persisted shape of an `object_type_bindings` row.
#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct ObjectTypeBindingRow {
    pub id: Uuid,
    pub object_type_id: Uuid,
    pub dataset_id: Uuid,
    pub dataset_branch: Option<String>,
    pub dataset_version: Option<i32>,
    pub primary_key_column: String,
    pub property_mapping: Value,
    pub sync_mode: String,
    pub default_marking: String,
    pub preview_limit: i32,
    pub owner_id: Uuid,
    pub last_materialized_at: Option<DateTime<Utc>>,
    pub last_run_status: Option<String>,
    pub last_run_summary: Option<Value>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Public API representation of a binding.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ObjectTypeBinding {
    pub id: Uuid,
    pub object_type_id: Uuid,
    pub dataset_id: Uuid,
    pub dataset_branch: Option<String>,
    pub dataset_version: Option<i32>,
    pub primary_key_column: String,
    pub property_mapping: Vec<ObjectTypeBindingPropertyMapping>,
    pub sync_mode: ObjectTypeBindingSyncMode,
    pub default_marking: String,
    pub preview_limit: i32,
    pub owner_id: Uuid,
    pub last_materialized_at: Option<DateTime<Utc>>,
    pub last_run_status: Option<String>,
    pub last_run_summary: Option<Value>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl TryFrom<ObjectTypeBindingRow> for ObjectTypeBinding {
    type Error = String;

    fn try_from(row: ObjectTypeBindingRow) -> Result<Self, Self::Error> {
        let property_mapping: Vec<ObjectTypeBindingPropertyMapping> =
            serde_json::from_value(row.property_mapping)
                .map_err(|error| format!("failed to decode property_mapping: {error}"))?;
        let sync_mode = ObjectTypeBindingSyncMode::try_from(row.sync_mode.as_str())?;
        Ok(Self {
            id: row.id,
            object_type_id: row.object_type_id,
            dataset_id: row.dataset_id,
            dataset_branch: row.dataset_branch,
            dataset_version: row.dataset_version,
            primary_key_column: row.primary_key_column,
            property_mapping,
            sync_mode,
            default_marking: row.default_marking,
            preview_limit: row.preview_limit,
            owner_id: row.owner_id,
            last_materialized_at: row.last_materialized_at,
            last_run_status: row.last_run_status,
            last_run_summary: row.last_run_summary,
            created_at: row.created_at,
            updated_at: row.updated_at,
        })
    }
}

#[derive(Debug, Deserialize)]
pub struct CreateObjectTypeBindingRequest {
    pub dataset_id: Uuid,
    pub dataset_branch: Option<String>,
    pub dataset_version: Option<i32>,
    pub primary_key_column: String,
    #[serde(default)]
    pub property_mapping: Vec<ObjectTypeBindingPropertyMapping>,
    pub sync_mode: ObjectTypeBindingSyncMode,
    pub default_marking: Option<String>,
    pub preview_limit: Option<i32>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateObjectTypeBindingRequest {
    pub dataset_branch: Option<String>,
    pub dataset_version: Option<i32>,
    pub primary_key_column: Option<String>,
    pub property_mapping: Option<Vec<ObjectTypeBindingPropertyMapping>>,
    pub sync_mode: Option<ObjectTypeBindingSyncMode>,
    pub default_marking: Option<String>,
    pub preview_limit: Option<i32>,
}

#[derive(Debug, Deserialize, Default)]
pub struct MaterializeBindingRequest {
    /// Optional override for the dataset branch to read.
    pub dataset_branch: Option<String>,
    /// Optional override for the dataset version to read.
    pub dataset_version: Option<i32>,
    /// Optional override for the per-run row limit. Capped at the binding's
    /// configured `preview_limit` for safety.
    pub limit: Option<i32>,
    /// When `true` no rows are written; only counts and validation errors are
    /// reported.
    #[serde(default)]
    pub dry_run: bool,
}

#[derive(Debug, Serialize)]
pub struct MaterializeBindingResponse {
    pub binding_id: Uuid,
    pub status: String,
    pub rows_read: i32,
    pub inserted: i32,
    pub updated: i32,
    pub skipped: i32,
    pub errors: i32,
    pub dry_run: bool,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub error_details: Vec<Value>,
}

#[derive(Debug, Serialize)]
pub struct ListObjectTypeBindingsResponse {
    pub data: Vec<ObjectTypeBinding>,
}

File: openfoundry-go/libs/ontology-kernel/models/object_type_binding.go
package models

import (
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"
)

// ObjectTypeBindingSyncMode mirrors `enum ObjectTypeBindingSyncMode`
// in `libs/ontology-kernel/src/models/object_type_binding.rs`.
// `#[serde(rename_all = "snake_case")]` and `#[sqlx(type_name =
// "text", rename_all = "snake_case")]` so wire / DB tokens are
// snake_case verbatim.
type ObjectTypeBindingSyncMode string

const (
	ObjectTypeBindingSyncModeSnapshot    ObjectTypeBindingSyncMode = "snapshot"
	ObjectTypeBindingSyncModeIncremental ObjectTypeBindingSyncMode = "incremental"
	ObjectTypeBindingSyncModeView        ObjectTypeBindingSyncMode = "view"
)

// AsStr mirrors `impl ObjectTypeBindingSyncMode::as_str(self)`.
func (m ObjectTypeBindingSyncMode) AsStr() string { return string(m) }

// ParseObjectTypeBindingSyncMode mirrors `TryFrom<&str>` — trims
// whitespace and rejects unknown tokens with the exact Rust error
// message format.
func ParseObjectTypeBindingSyncMode(s string) (ObjectTypeBindingSyncMode, error) {
	switch strings.TrimSpace(s) {
	case "snapshot":
		return ObjectTypeBindingSyncModeSnapshot, nil
	case "incremental":
		return ObjectTypeBindingSyncModeIncremental, nil
	case "view":
		return ObjectTypeBindingSyncModeView, nil
	default:
		return "", fmt.Errorf(
			"sync_mode '%s' is not supported; expected one of: snapshot, incremental, view",
			s,
		)
	}
}

// ObjectTypeBindingPropertyMapping mirrors `struct
// ObjectTypeBindingPropertyMapping`.
type ObjectTypeBindingPropertyMapping struct {
	SourceField    string `json:"source_field"`
	TargetProperty string `json:"target_property"`
}

// ObjectTypeBindingRow mirrors `struct ObjectTypeBindingRow` —
// persisted shape with raw JSONB columns kept as `json.RawMessage`.
type ObjectTypeBindingRow struct {
	ID                  uuid.UUID       `db:"id"`
	ObjectTypeID        uuid.UUID       `db:"object_type_id"`
	DatasetID           uuid.UUID       `db:"dataset_id"`
	DatasetBranch       *string         `db:"dataset_branch"`
	DatasetVersion      *int32          `db:"dataset_version"`
	PrimaryKeyColumn    string          `db:"primary_key_column"`
	PropertyMapping     json.RawMessage `db:"property_mapping"`
	SyncMode            string          `db:"sync_mode"`
	DefaultMarking      string          `db:"default_marking"`
	PreviewLimit        int32           `db:"preview_limit"`
	OwnerID             uuid.UUID       `db:"owner_id"`
	LastMaterializedAt  *time.Time      `db:"last_materialized_at"`
	LastRunStatus       *string         `db:"last_run_status"`
	LastRunSummary      json.RawMessage `db:"last_run_summary"`
	CreatedAt           time.Time       `db:"created_at"`
	UpdatedAt           time.Time       `db:"updated_at"`
}

// ObjectTypeBinding mirrors `struct ObjectTypeBinding` (public API
// representation).
type ObjectTypeBinding struct {
	ID                 uuid.UUID                          `json:"id"`
	ObjectTypeID       uuid.UUID                          `json:"object_type_id"`
	DatasetID          uuid.UUID                          `json:"dataset_id"`
	DatasetBranch      *string                            `json:"dataset_branch"`
	DatasetVersion     *int32                             `json:"dataset_version"`
	PrimaryKeyColumn   string                             `json:"primary_key_column"`
	PropertyMapping    []ObjectTypeBindingPropertyMapping `json:"property_mapping"`
	SyncMode           ObjectTypeBindingSyncMode          `json:"sync_mode"`
	DefaultMarking     string                             `json:"default_marking"`
	PreviewLimit       int32                              `json:"preview_limit"`
	OwnerID            uuid.UUID                          `json:"owner_id"`
	LastMaterializedAt *time.Time                         `json:"last_materialized_at"`
	LastRunStatus      *string                            `json:"last_run_status"`
	LastRunSummary     json.RawMessage                    `json:"last_run_summary"`
	CreatedAt          time.Time                          `json:"created_at"`
	UpdatedAt          time.Time                          `json:"updated_at"`
}

// IntoBinding mirrors `TryFrom<ObjectTypeBindingRow> for
// ObjectTypeBinding`. Returns the embedded JSON parse error or
// sync_mode validation error verbatim.
func (row ObjectTypeBindingRow) IntoBinding() (ObjectTypeBinding, error) {
	mapping := []ObjectTypeBindingPropertyMapping{}
	if len(row.PropertyMapping) > 0 {
		if err := json.Unmarshal(row.PropertyMapping, &mapping); err != nil {
			return ObjectTypeBinding{}, fmt.Errorf("failed to decode property_mapping: %s", err)
		}
	}
	mode, err := ParseObjectTypeBindingSyncMode(row.SyncMode)
	if err != nil {
		return ObjectTypeBinding{}, err
	}
	return ObjectTypeBinding{
		ID:                 row.ID,
		ObjectTypeID:       row.ObjectTypeID,
		DatasetID:          row.DatasetID,
		DatasetBranch:      row.DatasetBranch,
		DatasetVersion:     row.DatasetVersion,
		PrimaryKeyColumn:   row.PrimaryKeyColumn,
		PropertyMapping:    mapping,
		SyncMode:           mode,
		DefaultMarking:     row.DefaultMarking,
		PreviewLimit:       row.PreviewLimit,
		OwnerID:            row.OwnerID,
		LastMaterializedAt: row.LastMaterializedAt,
		LastRunStatus:      row.LastRunStatus,
		LastRunSummary:     row.LastRunSummary,
		CreatedAt:          row.CreatedAt,
		UpdatedAt:          row.UpdatedAt,
	}, nil
}

// CreateObjectTypeBindingRequest mirrors the same struct in Rust.
// `property_mapping` is `#[serde(default)]` so missing key decodes
// to `[]`.
type CreateObjectTypeBindingRequest struct {
	DatasetID        uuid.UUID                          `json:"dataset_id"`
	DatasetBranch    *string                            `json:"dataset_branch,omitempty"`
	DatasetVersion   *int32                             `json:"dataset_version,omitempty"`
	PrimaryKeyColumn string                             `json:"primary_key_column"`
	PropertyMapping  []ObjectTypeBindingPropertyMapping `json:"property_mapping"`
	SyncMode         ObjectTypeBindingSyncMode          `json:"sync_mode"`
	DefaultMarking   *string                            `json:"default_marking,omitempty"`
	PreviewLimit     *int32                             `json:"preview_limit,omitempty"`
}

// UnmarshalJSON applies the `#[serde(default)]` for property_mapping.
func (r *CreateObjectTypeBindingRequest) UnmarshalJSON(b []byte) error {
	type alias CreateObjectTypeBindingRequest
	if err := json.Unmarshal(b, (*alias)(r)); err != nil {
		return err
	}
	if r.PropertyMapping == nil {
		r.PropertyMapping = []ObjectTypeBindingPropertyMapping{}
	}
	return nil
}

// UpdateObjectTypeBindingRequest mirrors `struct UpdateObjectTypeBindingRequest`.
type UpdateObjectTypeBindingRequest struct {
	DatasetBranch    *string                              `json:"dataset_branch,omitempty"`
	DatasetVersion   *int32                               `json:"dataset_version,omitempty"`
	PrimaryKeyColumn *string                              `json:"primary_key_column,omitempty"`
	PropertyMapping  *[]ObjectTypeBindingPropertyMapping  `json:"property_mapping,omitempty"`
	SyncMode         *ObjectTypeBindingSyncMode           `json:"sync_mode,omitempty"`
	DefaultMarking   *string                              `json:"default_marking,omitempty"`
	PreviewLimit     *int32                               `json:"preview_limit,omitempty"`
}

// MaterializeBindingRequest mirrors `struct MaterializeBindingRequest`.
// `dry_run` is `#[serde(default)]` so absent → false.
type MaterializeBindingRequest struct {
	DatasetBranch  *string `json:"dataset_branch,omitempty"`
	DatasetVersion *int32  `json:"dataset_version,omitempty"`
	Limit          *int32  `json:"limit,omitempty"`
	DryRun         bool    `json:"dry_run"`
}

// MaterializeBindingResponse mirrors `struct MaterializeBindingResponse`.
// `error_details` is `#[serde(skip_serializing_if = "Vec::is_empty")]`
// — `omitempty` honours that for `[]json.RawMessage` zero length.
type MaterializeBindingResponse struct {
	BindingID    uuid.UUID         `json:"binding_id"`
	Status       string            `json:"status"`
	RowsRead     int32             `json:"rows_read"`
	Inserted     int32             `json:"inserted"`
	Updated      int32             `json:"updated"`
	Skipped      int32             `json:"skipped"`
	Errors       int32             `json:"errors"`
	DryRun       bool              `json:"dry_run"`
	ErrorDetails []json.RawMessage `json:"error_details,omitempty"`
}

// ListObjectTypeBindingsResponse mirrors `struct ListObjectTypeBindingsResponse`.
type ListObjectTypeBindingsResponse struct {
	Data []ObjectTypeBinding `json:"data"`
}
